From 4686468ccf301b2309ba323e29da0948d9bff872 Mon Sep 17 00:00:00 2001
From: Sandor Yu <Sandor.yu@nxp.com>
Date: Wed, 24 Jan 2018 19:18:16 +0800
Subject: [PATCH 3253/5242] MLK-17456-1: hdp: Enable EDID read function for
 imx8mq hdmi

commit  c9027d64e18f4bada68f1ad68b4ad5bbc4afbf66 from
https://source.codeaurora.org/external/imx/linux-imx.git

-Enable EDID read function for imx8mq hdmi.
-Add video mode check.

Signed-off-by: Sandor Yu <Sandor.yu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/hdp/imx-hdp.c |   18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/imx/hdp/imx-hdp.c b/drivers/gpu/drm/imx/hdp/imx-hdp.c
index ff856a0..fedd775 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdp.c
+++ b/drivers/gpu/drm/imx/hdp/imx-hdp.c
@@ -670,14 +670,12 @@ static void imx_hdp_bridge_enable(struct drm_bridge *bridge)
 static int imx_hdp_connector_get_modes(struct drm_connector *connector)
 {
 	struct drm_display_mode *mode;
+	struct imx_hdp *hdp = container_of(connector, struct imx_hdp, connector);
+	struct edid *edid;
 	int num_modes = 0;
 	int ret;
 	int i;
 
-	struct imx_hdp *hdp = container_of(connector, struct imx_hdp,
-					     connector);
-	struct edid *edid;
-
 	if (hdp->is_edid == true) {
 		edid = drm_do_get_edid(connector, hdp->ops->get_edid_block, &hdp->state);
 		if (edid) {
@@ -714,13 +712,21 @@ static int imx_hdp_connector_get_modes(struct drm_connector *connector)
 					     connector);
 	enum drm_mode_status mode_status = MODE_OK;
 	struct drm_cmdline_mode *cmdline_mode;
+	u32 hdp_vic;
+
 	cmdline_mode = &connector->cmdline_mode;
 
 	/* cmdline mode is the max support video mode when edid disabled */
-	if (!hdp->is_edid)
+	if (!hdp->is_edid) {
 		if (cmdline_mode->xres != 0 &&
 			cmdline_mode->xres < mode->hdisplay)
 			return MODE_BAD_HVALUE;
+	} else {
+		/* Check mode in hdp vic table */
+		hdp_vic = imx_get_vic_index(mode);
+		if (hdp_vic < 0)
+			return MODE_NOMODE;
+	}
 
 	if (hdp->is_4kp60 && mode->clock > 594000)
 		return MODE_CLOCK_HIGH;
@@ -966,7 +972,7 @@ static int imx8qm_hdp_swrite(struct hdp_mem *mem, unsigned int addr, unsigned in
 };
 
 static struct hdp_devtype imx8mq_hdmi_devtype = {
-	.is_edid = false,
+	.is_edid = true,
 	.is_4kp60 = true,
 	.audio_type = CDN_HDMITX_KIRAN,
 	.ops = &imx8mq_ops,
-- 
1.7.9.5

